subprojects/persistent-cache/src/main/java/org/gradle/cache/internal/ProducerGuard.java

```diff
@@ -65,6 +70,18 @@
         return new SerialProducerGuard<T>();
     }
 
+    /**
+     * Creates a {@link ProducerGuard} which guarantees that different keys will block on different locks,
+     * ensuring maximum concurrency, but will also timeout when the lock cannot be acquired.
+     * This can be used typically when debugging deadlocks. Please notice however that this kind of guard
+     * consumes significantly more resources than the {@link #striped() striped} and {@link #adaptive() adaptive}
+     * versions.
+     * If the timeout is reached, a timeout exception will be thrown.
+     */
+    public static <T> ProducerGuard<T> timingOut(int value, TimeUnit timeUnit) {
```

Comment:
Is this for debugging? I guess we do not use it in production, right?

Reply:
We don't. The javadocs say it's intended for debugging, but there may be situations in the future where it makes sense to use them.